### PR TITLE
[performance] logql：Experimental feature Optimize json parser in limited cases.

### LINKED ISSUE #5405: [performance] logql：json cost 64 % s

### PR DESCRIPTION
**Pull request by liguozhong**
**What this PR does / why we need it**:
Experimental feature Optimize json parser in limited cases.
Implement a high-performance simplejson parser（labels_hints_fast_parse）.
```shell
Benchmark_JsonParserFastParse/json/no_labels_hints-12         	  272154	      4296 ns/op
Benchmark_JsonParserFastParse/json/labels_hints_fast_parse-12 	 4061995	       258.0 ns/op
```
for example:
```go
logql ：
{functionName="feedflow"}  | json   | durationMs > 2003 
transform to
Optimize logql ：
{ functionName="feedflow"} | simplejson | durationMs>2003 | json 
```

In this way, reduce the performance loss caused by full json key parsing

**Which issue(s) this PR fixes**:
Fixes #5405>

**Special notes for your reviewer**:
"linefmt" case logql is more complicated, and another PR will handle it.
This feature is turned off by default, I will turn on this switch in our loki cluster to continuously test and trace this feature

loki.yaml config
```yaml
querier:
  query_timeout: 30m
  engine:
    logql_optimize_enable: true
  query_ingesters_within: 2h
```

**online environment**
I have completed the test in the online environment
tracing
image
+ performance = 50s-39/50s = 22%.  I will continue profile this json issue.

with this pr 39s
image
without this pr 50s
image
**Checklist**
- [x] Have you completed the test in the online environment
- [x] Documentation added
- [x] Tests updated
- [x] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

**Comment by splitice** [comment]
Isnt the test a no-op?

`{log_type="service_metrics", module="api_server", operation="InvokeFunction", accountID="212068714932184585", serviceName="taojimu-fc-prod", functionName="feedflow"} | json | durationMs>2003 | json`

Or is this some new magic?

File: pkg/logql/engine.go [hunk 1]
```diff
@@ -43,11 +44,15 @@ type EngineOpts struct {
...
 	// only used for instant log queries.
 	MaxLookBackPeriod time.Duration `yaml:"max_look_back_period"`
+
+	LogqlOptimizeEnable bool `yaml:"logql_optimize_enable"`
 }
 
 func (opts *EngineOpts) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
 	f.DurationVar(&opts.Timeout, prefix+".engine.timeout", 5*time.Minute, "Timeout for query execution.")
 	f.DurationVar(&opts.MaxLookBackPeriod, prefix+".engine.max-lookback-period", 30*time.Second, "The maximum amount of time to look back for log lines. Used only for instant log queries.")
+	f.BoolVar(&opts.LogqlOptimizeEnable, "logql-optimize-enablelog-config-reverse-order", false, "Experimental feature logql optimization")
```
Comment (kavirajk) [inline pkg/logql/engine.go:54]:
```suggestion
	f.BoolVar(&opts.LogqlOptimizeEnable, prefix +".engine.logql-optimize-enable", false, "Experimental feature logql optimization")
```
Can we have same convention for flag like others?

File: pkg/logql/log/parser.go [hunk 2]
```diff
@@ -39,6 +40,8 @@ type JSONParser struct {
 	lbs *LabelsBuilder
 
 	keys internedStringSet
+
+	RequiredJsonLabels []string
```
Comment (kavirajk) [inline pkg/logql/log/parser.go:44]: Q: So this `RequiredJsonLabels` are actually taken from the logql expression based on comes after `| json` filter correct?
 e.g: ` ... | json | foo="bar" `

Here the required labels are just `foo`? Is my understanding right?

Comment (liguozhong) [reply pkg/logql/log/parser.go:44]: > 
Completely correct！
I added a test function to make the RequiredJsonLabels logic easier to understand.

image

**Comment by liguozhong** [comment]
> Isnt the test a no-op?
> 
> `{log_type="service_metrics", module="api_server", operation="InvokeFunction", accountID="212068714932184585", serviceName="taojimu-fc-prod", functionName="feedflow"} | json | durationMs>2003 | json`
> 
> Or is this some new magic?

I added a test function to make the RequiredJsonLabels logic easier to understand.

File: pkg/logql/optimize_test.go [hunk 3]
```diff
@@ -35,3 +36,30 @@ func Test_optimizeSampleExpr(t *testing.T) {
...
+
+func Test_optimizeLogSelectorExpr(t *testing.T) {
+	tests := []struct {
+		name, logql, expected string
+	}{
+		// noop
+		{
+			"my slow json parser case ",
+			`{log_type="service_metrics",module="api_server",operation="InvokeFunction",accountID="212068714932184585",serviceName="taojimu-fc-prod",functionName="feedflow"}  | json   | durationMs > 2003`,
+			`{log_type="service_metrics", module="api_server", operation="InvokeFunction", accountID="212068714932184585", serviceName="taojimu-fc-prod", functionName="feedflow"} | json | durationMs>2003 | json`,
```
[the remark this answers is not in the document]
Comment (liguozhong) [reply pkg/logql/optimize_test.go:52]: | json | durationMs>2003 | json
==
| json（Hint：durationMs key） | durationMs>2003 | json（Hint：None）

**Comment by liguozhong** [comment]
> Isnt the test a no-op?
> 
> `{log_type="service_metrics", module="api_server", operation="InvokeFunction", accountID="212068714932184585", serviceName="taojimu-fc-prod", functionName="feedflow"} | json | durationMs>2003 | json`
> 
> Or is this some new magic?

| json | durationMs>2003 | json
～=
| simpleJson） | durationMs>2003 | json
～=
| json（Hint：durationMs key） | durationMs>2003 | json（Hint：None）

File: pkg/logql/log/parser.go [hunk 4]
```diff
@@ -49,10 +53,30 @@ func NewJSONParser() *JSONParser {
 	}
 }
 
+func (j *JSONParser) SetHint(hint string) {
```
Comment (jeschkies) [inline pkg/logql/log/parser.go:56]: What is a hint and why is it improving performance? It's not quite clear to me.

Comment (liguozhong) [reply pkg/logql/log/parser.go:56]: for example：
logql：
{functionName="feedflow"} | json  | durationMs > 2003
hint is (durationMs> 2003)

simpleJsonParser = (log [ IndexOf(log,"durationMs") , "," ] )

indexOf func is more performant than json func
log
`json
{"level":"info","message":"metrics","functionCostMBMs":193536,"resolveQualifierLatency":0.007786,"isCustomDomainInvoke":true,"serviceLatency":5.307937999999922,"metaLatency":0.001109,"cause":"disableDragonGlobal","scheduleTrace":"0:C2J9JDkvZB18mMZDxAGvhCNjYjQ1MzRiMy1iNDEyLTRmOTYtYWYzNi1hOGVmZDFjYTU1ZWYjZmVlZGZsb3dfX3N0YWJsZSM1ZjFmMWZiYS0yYmQwLTQ5NzAtODA1OS0yNDUzOWVlZmU3ZmEjMDA=","externalServiceVersion":"101","accessKeyID":"","userAgent":"ReactorNetty/0.9.18.RELEASE","numErrors":0,"pickBusyLimit":0,"concurrentCount":27,"functionErrors":0,"durationMs":2099.121799,"forwardedFor":"","timestamp":1644991121,"partitionCount":1,"eeAddress":"33.40.79.69","isHTTPTrigger":false,"requestID":"998b00b1-cd87-4244-862e-f763b0010ebc","fcStatus":200,"concurrentInvocationsPerFunction":27,"functionName":"feedflow","meteringDuration":62846814,"serviceName":"taojimu-fc-prod","traceID":"","method":"POST","network":"none","codeChecksum":"14483028719241002601","eeid":"faas-biz--ecb6be6f-e4d1-4508-a161-1c558348957c_20220216054348.718","mfaPresent":"","preIdleContainer":false,"containerID":"e9d41abe-2215-494a-a474-bdffd705cf8e","invokeDuration":2099121799,"invokeID":"0fb74b04-48ee-46ec-ae24-936939c4201d","queryPathLatency":0.001546,"ip":"33.45.125.83","requestOriginalURI":"/2016-08-15/services/taojimu-fc-prod.feedflow__stable/functions/feedflow/invocations","concurrentInvocations":27,"qualifier":"feedflow__stable","scheduleLatency":0.552185,"maxMemoryUsage":374.21484375,"operation":"InvokeFunction","apiVersion":"2016-08-15","containerReqCount":10,"metaCodeChecksum":"14483028719241002601","host":"feedflow__stable.feedflow.taojimu-fc","errorCode":"","hybridFallBack":true,"codeVersion":"","eaLatency":2103.270315,"concurrentReqLimit":300,"region":"cn-shanghai-corp","module":"api_server","hasFCError":false,"idleFlag":0,"eagleeyeTraceID":"21207bfa16449911210962677efeeb","meteredRespSize":40555,"timestampH":"2022-02-16 05:58:41.100","memoryLimitInMB":3072,"functionTimeoutInSec":50,"runtime":"alinode5","idleLatency":0.474576,"concurrencySoftLimit":0,"enableSlowClientOpt":false,"mode":"sync","accountID":"212068714932184585","isHTTPS":false,"concurrentReqLimitPerFunction":50,"isDarkLaunch":false,"clusterName":"fc-sh-jituan","status":200,"scheduleRetryCount":1,"metaLastModified":"2021-12-21T02:19:39Z","resourceMode":4,"hostName":"function-compute033042074246.ea119","path":"/2016-08-15/services/taojimu-fc-prod.feedflow__stable/functions/feedflow/invocations","latency":2104.429737,"concurrencyLimit":10,"instanceType":"ElasticInstance","invokeException":""}
`

File: pkg/logql/log/parser.go [hunk 5]
```diff
@@ -164,6 +188,49 @@ func (j *JSONParser) parseLabelValue(iter *jsoniter.Iterator, prefix, field stri
...
+			// case:string
+			valStr := value[1 : valLen-1]
+			lbs.Set(requiredKey, valStr)
+		} else {
+			lbs.Set(requiredKey, value)
+		}
+	}
+
+	return line, true
+}
```
Comment (cyriltovena) [inline pkg/logql/log/parser.go:232]: There's some json library that I've been looking at that does this in a more clean way.

For example the each object of https://github.com/buger/jsonparser

**Review by cyriltovena — CHANGES_REQUESTED** [review]
I support this idea, I think this could be better specially using hints, but I disagree with the implementation I think this should be the default path not something experimental or a replacement of LogQL queries.

My suggestion is to look into https://github.com/buger/jsonparser or even simdjson and replace the current parser which does an allocation for each key at minimum.

But I don't think a custom parsing will do it, there's so many special case to handle like quoting, escaping, etc....

File: pkg/logql/log/parser.go (hunk 5 above)
Comment (cyriltovena) [reply pkg/logql/log/parser.go:232]: see https://github.com/grafana/loki/issues/4328

File: pkg/logql/log/parser.go (hunk 4 above)
Comment (jeschkies) [reply pkg/logql/log/parser.go:56]: Ah. So it's looking ahead. I wonder if a parser Cyril mentioned below plus lazy parsing would be cleaner to implement.

**Comment by liguozhong** [comment]
> I support this idea, I think this could be better specially using hints, but I disagree with the implementation I think this should be the default path not something experimental or a replacement of LogQL queries.
> 
> My suggestion is to look into https://github.com/buger/jsonparser or even simdjson and replace the current parser which does an allocation for each key at minimum.
> 
> But I don't think a custom parsing will do it, there's so many special case to handle like quoting, escaping, etc....

Thanks for your review。
 I was busy with the coding test of another PR a few days ago (https://github.com/grafana/loki/pull/5455). 

Now that the test has been completed, I will start to see how the json library you suggested is used ，and try to optimize this PR

File: pkg/logql/log/parser.go (hunk 4 above)
Comment (liguozhong) [reply pkg/logql/log/parser.go:56]: Thanks for your review。
I was busy with the coding test of another PR a few days ago (https://github.com/grafana/loki/pull/5455).

Now that PR has been completed, I will try to optimize this PR

**Comment by cyriltovena** [comment]
Appreciate the work, and I do think it has value since I created an issue. Just skeptic about the current direction.

**Comment by liguozhong** [comment]
> Appreciate the work, and I do think it has value since I created an issue. Just skeptic about the current direction.

https://github.com/grafana/loki/pull/5538, I submit another PR

**Pull request closed without merging**